src/Services/ElementTabProvider.php

```diff
@@ -0,0 +1,123 @@
+<?php
+namespace DNADesign\Elemental\Services;
+
+use DNADesign\Elemental\Models\BaseElement;
+use Psr\SimpleCache\CacheInterface;
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Flushable;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Forms\TabSet;
+
+/**
+ * Provides tab names for
+ */
+class ElementTabProvider implements Flushable
+{
+    use Configurable;
+
+    /**
+     * Set to true to refresh the tab cache on flush
+     *
+     * @config
+     * @var bool
+     */
+    private static $regenerate_on_flush = false;
+
+    /**
+     * @var CacheInterface
+     */
+    protected $cache;
+
+    /**
+     * ElementTabProvider constructor.
+     * @param CacheInterface $cache
+     */
+    public function __construct(CacheInterface $cache = null)
+    {
+        if (!$cache) {
+            $cache = Injector::inst()->get(CacheInterface::class . '.ElementTabCache');
+        }
+        $this->cache = $cache;
+    }
+
+    /**
+     * Get the top level tab names for the given element class
+     *
+     * @param string $elementClass
+     * @return array Array of the tabs for the element
+     */
+    public function getTabsForElement($elementClass)
+    {
+        if (null !== ($tabs = $this->cache->get($this->getCacheKey($elementClass)))) {
+            return $tabs;
+        }
+
+        return $this->generateTabsForElement($elementClass);
+    }
+
+    /**
+     * This function is triggered early in the request if the "flush" query
+     * parameter has been set. Each class that implements Flushable implements
+     * this function which looks after it's own specific flushing functionality.
+     *
+     * @see FlushMiddleware
+     */
+    public static function flush()
+    {
+        /** @var static $self */
+        $self = singleton(static::class);
+
+        $self->cache->clear();
+
+        if (self::config()->regenerate_on_flush) {
```

Comment:
It might be preference, but I'd rather see it written like this:

```diff
- self::config()->regenerate_on_flush
+ $this->config()->get('regenerate_on_flush')
```

Reply:
Okay I'll update. My preference is still to use `self` I'd say. `config()` is marked as static and `self` is more descriptive about where the config is coming from.

Reply:
Also in this case it's a static method - so I updated it to `$self->config()->get()`.

Reply:
Ah sorry I missed that. That's fine. I do think it's a bit nicer this way since the config is obviously part of the same instance, even though in practice it wouldn't matter. Cheers